### DISCUSSION TITLE
API: Query IP Addresses by Interface

### DISCUSSION THREAD

**Discussion by AnneWielis** [netbox-community/netbox#6489] [Q&A] [answered]
When i try to get the IP of an interface like this:

http://netbox/api/ipam/ip-addresses/?assigned_object_id=1533

I get back all the IPs in the System. 

What am i doing wrong?

**Comment by candlerb** [comment, marked as answer]
If you give an unrecognised query parameter then it's ignored (without any error), which is why you get them all.

I agree it's not well documented, but the defined set of filters for IPAddress objects is [here](https://github.com/netbox-community/netbox/blob/develop/netbox/ipam/filtersets.py#L373-L467).

The ones you want are "interface_id" or "vminterface_id".  That is: if 1533 is an interface on a device, then use `http://netbox/api/ipam/ip-addresses/?interface_id=1533`.  If it's on a VM, then use `http://netbox/api/ipam/ip-addresses/?vminterface_id=1533`

**Reply by candlerb** [reply]
Actually, you can get this information from the web interface.

* Open the Netbox front page
* Click on "API" link at the bottom, takes you to `/api/docs/`
* Scroll down to `GET /ipam/ip-addresses/`
* Click on it to expand
* Look through the "parameters" section (pale blue box), it includes:

![image](https://user-images.githubusercontent.com/44789/119800389-bee5af80-bed4-11eb-9f8e-b2e26ee08b43.png)

**Comment by AnneWielis** [comment]
OK, i see...so i can't search for virtually just everything, but the allowed queries are hardcoded...OK, got that now.

Thanks for pointing out!

**Comment by AnneWielis** [comment]
That brings me to the point, why i can't search for description? That does work on interfaces, though, and i abuse that as foreign key for importing stuff from somewhere else

**Reply by candlerb** [reply]
An API search on description works for me:

```
$ curl -Ss http://netbox/api/ipam/ip-addresses/?description=ganeti+cluster+master | jq . | grep description
      "description": "ganeti cluster master",
$
```

It's an exact match rather than partial, so you need to give the actual value.  If you want partial matching use `q` instead, although that [matches](https://github.com/netbox-community/netbox/blob/develop/netbox/ipam/filtersets.py#L477-L479) dns name, description, and IP address

```
$ curl -Ss http://netbox/api/ipam/ip-addresses/?q=cluster | jq . | grep description
      "description": "ganeti cluster master",
$ 
```

For your use case, I think that configuring a custom field would be better than abusing description.  I believe you should be able to search on them using `cf_<fieldname>` as the query parameter.

**Reply by AnneWielis** [reply]
Hm, i tried: http://netbox/api/ipam/ip-addresses/?description=999, that didn't work, but http://netbox/api/ipam/ip-addresses/?q=999 did. Weird...

**Reply by candlerb** [reply]
I just tried adding a description "999" to an IP address, then searching for it with `/api/ipam/ip-addresses/?description=999`, and it worked fine.  You sure you don't have a space in there or something?

**Reply by candlerb** [reply]
I don't believe you said what version of Netbox you are using.  I am using the most recent release, v2.11.4, and that includes a fix mentioned in the [release notes](https://github.com/netbox-community/netbox/releases/tag/v2.11.4):

* #6393 - Add `description` filter for IP addresses

So it won't work if you're on v2.11.3 or earlier.

**Reply by AnneWielis** [reply]
You are right, 2.11.4 is just 2 days old...installed it now. Works fine, thanks!
